### DISCUSSION TITLE
why can't stream() be used with `isolation_level='AUTOCOMMIT'` ? (postgresql)

### DISCUSSION THREAD

**Discussion by wolph** [sqlalchemy/sqlalchemy#8930] [Usage Questions]
### Describe the bug

When the `isolation_level` for the engine is set  to `AUTOCOMMIT` it breaks the `stream()` function because no transaction is started.

### To Reproduce

```python
import asyncio

import sqlalchemy
from sqlalchemy.ext.asyncio import async_sessionmaker, create_async_engine


async def main():
    engine = create_async_engine(
        'postgresql+psycopg://',
        isolation_level='AUTOCOMMIT',
    )
    async with engine.begin() as conn:
        statement = sqlalchemy.text('SELECT 1 UNION ALL SELECT 2')
        async for row in await conn.stream(statement):
            print(row)

    await engine.dispose()


asyncio.run(main())
```

### Error

```
Traceback (most recent call last):
  File "sqlalchemy/engine/base.py", line 1964, in _exec_single_context
    self.dialect.do_execute(
  File "sqlalchemy/engine/default.py", line 743, in do_execute
    cursor.execute(statement, parameters)
  File "sqlalchemy/dialects/postgresql/psycopg.py", line 585, in execute
    self.await_(self._cursor.execute(query, params, **kw))
  File "sqlalchemy/util/_concurrency_py3k.py", line 102, in await_only
    return current.driver.switch(awaitable)  # type: ignore[no-any-return]
  File "sqlalchemy/util/_concurrency_py3k.py", line 160, in greenlet_spawn
    value = await result
  File "psycopg/server_cursor.py", line 417, in execute
    raise ex.with_traceback(None)
psycopg.errors.NoActiveSqlTransaction: DECLARE CURSOR can only be used in transaction blocks

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test.py", line 25, in <module>
    asyncio.run(main())
  File "asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "test.py", line 19, in main
    async for row in await conn.stream(statement):
  File "sqlalchemy/ext/asyncio/base.py", line 170, in start
    start_value = await util.anext_(self.gen)
  File "sqlalchemy/ext/asyncio/engine.py", line 517, in stream
    result = await greenlet_spawn(
  File "sqlalchemy/util/_concurrency_py3k.py", line 167, in greenlet_spawn
    result = context.switch(value)
  File "sqlalchemy/engine/base.py", line 1414, in execute
    return meth(
  File "sqlalchemy/sql/elements.py", line 487, in _execute_on_connection
    return connection._execute_clauseelement(
  File "sqlalchemy/engine/base.py", line 1638, in _execute_clauseelement
    ret = self._execute_context(
  File "sqlalchemy/engine/base.py", line 1842, in _execute_context
    return self._exec_single_context(
  File "sqlalchemy/engine/base.py", line 1983, in _exec_single_context
    self._handle_dbapi_exception(
  File "sqlalchemy/engine/base.py", line 2325, in _handle_dbapi_exception
    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
  File "sqlalchemy/engine/base.py", line 1964, in _exec_single_context
    self.dialect.do_execute(
  File "sqlalchemy/engine/default.py", line 743, in do_execute
    cursor.execute(statement, parameters)
  File "sqlalchemy/dialects/postgresql/psycopg.py", line 585, in execute
    self.await_(self._cursor.execute(query, params, **kw))
  File "sqlalchemy/util/_concurrency_py3k.py", line 102, in await_only
    return current.driver.switch(awaitable)  # type: ignore[no-any-return]
  File "sqlalchemy/util/_concurrency_py3k.py", line 160, in greenlet_spawn
    value = await result
  File "psycopg/server_cursor.py", line 417, in execute
    raise ex.with_traceback(None)
sqlalchemy.exc.InternalError: (psycopg.errors.NoActiveSqlTransaction) DECLARE CURSOR can only be used in transaction blocks
[SQL: SELECT 1 UNION ALL SELECT 2]
(Background on this error at: https://sqlalche.me/e/20/2j85)
``` 

### Versions

- OS: OSX 12.6.1 and Windows 10
- Python: 3.10
- SQLAlchemy: 2.0.0b3
- Database: PostgreSQL 15
- DBAPI psycopg

**Comment by wolph** [comment]
It shouldn't be a discussion, it's definitely a bug. Running `await session.begin()` before doesn't work either. Only `session.execute(sqlalchemy.text('BEGIN;'))` will make it "work"

**Reply by zzzeek** [reply]
.begin() is no-op in SQLAlchemy as far as emitting any commands to the database, regardless of whether or not AUTOCOMMIT is set.  The DBAPI has only ``.rollback()`` and ``.commit()`` methods, not ``.begin()``.

**Reply by zzzeek** [reply]
also even if there were a begin() method, SQLAlchemy would never invoke it implicitly within autocommit mode, that would be silently turning autocommit mode off.

**Reply by wolph** [reply]
It might be useful to have a big fat error when executing stream in autocommit mode in that case :) It took me quite a bit of time to hunt down the cause of this issue

**Reply by zzzeek** [reply]
the error that PostgreSQL raised IMO should be pretty sufficient....I dont think most people expect AUTOCOMMIT mode to be silently suspended...

**Comment by zzzeek** [comment]
PostgreSQL's DECLARE CURSOR statement is only supported when transactions are in progress.   By definition, SQLAlchemy's AUTOCOMMIT mode does not support transactions; SQLAlchemy's transaction management relies upon the DBAPI to automatically BEGIN transactions which does not occur during AUTOCOMMIT.  

The .autocommit attribute in psycopg (described at https://www.psycopg.org/docs/connection.html?highlight=autocommit#connection.autocommit) is similar to the .autocommit attribute in many other DBAPIs, all of which act this way.  As such, a formal definition of .autocommit was just added to pep-249 at https://www.psycopg.org/docs/connection.html?highlight=autocommit#connection.autocommit:


> Attribute to query and set the autocommit mode of the connection.

> Return True if the connection is operating in autocommit (**non- transactional**) mode. Return False if the connection is operating in manual commit (**transactional**) mode. 
 
Therefore to use server side cursors with PostgreSQL you would need to use a connection that does not have the "AUTOCOMMIT" isolation level set.

**Reply by wolph** [reply]
Ah... that's completely not as I expected.

Within a regular PostgreSQL database session (i.e. using `psql`) you can still manually start a transaction if you need control about the transaction so I was not expecting this behaviour.

Due the `idle in transaction` scenario with PostgreSQL it's often beneficial to have every connection running in autocommit mode. When you need an explicit transaction, it should still be possible to create a transaction which will temporarily (until rollback/commit) "disable" autocommit

**Reply by zzzeek** [reply]
> Ah... that's completely not as I expected.
> 
> Within a regular PostgreSQL database session (i.e. using `psql`) you can still manually start a transaction if you need control about the transaction so I was not expecting this behaviour.

There's also some docs about how SQLAlchemy interprets begin() at https://docs.sqlalchemy.org/en/14/core/connections.html#understanding-the-dbapi-level-autocommit-isolation-level .  I thought we had more mentions of this, we should at least have a note at https://docs.sqlalchemy.org/en/14/core/connections.html#using-transactions that pep-249 DBAPI has no begin() method.

You can also look at the behavior of psycopg or psycopg2 itself.   You get a connection, set .autocommit on it, then try to get a cursor() with a server-side name, that throws this error.  The DBAPI, which is normally what emits BEGIN automatically, doesn't do it in this case.

> Due the `idle in transaction` scenario with PostgreSQL it's often beneficial to have every connection running in autocommit mode. 

That should not be necessary if the application releases transaction and connection resources when it's done with an operation, and similarly doesn't do any long running tasks without first ensuring that the current transaction is rolled back or committed.

".autocommit" mode is mostly beneficial for the MySQL database where transactions being present tends to have a significant performance penalty.

> When you need an explicit transaction, it should still be possible to create a transaction which will temporarily (until rollback/commit) "disable" autocommit

this is not the purpose of autocommit mode in the DBAPI.  if you need a transcation, turn off autocommit.

**Reply by wolph** [reply]
I guess I just misinterpreted the documentation here: https://docs.sqlalchemy.org/en/20/orm/session_transaction.html
Specifically: "such that the [Session.begin()](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.begin) method may be called directly for a [Session](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session) that has not already had a transaction begun"
